tests/handlers/test_federation.py

```diff
@@ -314,142 +312,6 @@ def test_backfill_with_many_backward_extremities(self) -> None:
             )
         self.get_success(d)
 
-    def test_backfill_floating_outlier_membership_auth(self) -> None:
```

Comment:
This test was added in #10439, but I don't understand why it is correct (and it now fails). In particular, it seems to assert that a user that is not a member of the room according to our idea of the room state should be able to send an event, which seems wrong, and in conflict with point 5 at https://spec.matrix.org/v1.3/server-server-api/#checks-performed-on-receipt-of-a-pdu:

> 5. Passes authorization rules based on the state before the event, otherwise it is rejected.

cc @MadLittleMods

Reply:
I've stared at this for a while and I my thoughts have boiled down to this (probably missing something obvious).

 - The member event is authed by the previous state events in the room (auth events defined).
 - The new message is authed by the member and and previous state events in the room.
 - `prev_events` don't matter. It's all about `auth_events`. At least that's how it has been working in Synapse for me while playing with the MSC2716 stuff.
 - I don't know how everything should ideally work according to the spec (state res hard).



---

For my own reference,

 - The full changes from [#10439](https://github.com/matrix-org/synapse/pull/10439) are visible in https://github.com/matrix-org/synapse/pull/10439/files/eb32018cb3d9e048fa4b3523a6fb29b1e2b36faf (some changes hidden by changes being merged in via [#10245](https://github.com/matrix-org/synapse/pull/10245) and `develop` being merged).
 - Relevant outdated diff comment, https://github.com/matrix-org/synapse/pull/10439#discussion_r673820000
 - Following where the `missing_auth_event_context` fix from that diff went:
    - Moved to `federation_event.py` in https://github.com/matrix-org/synapse/pull/10692/files#diff-a91beeaf44154c199a11f5539f2358704a1df966ff74ad8f2681b64430843f8cR1527
    - Removed in https://github.com/matrix-org/synapse/pull/10896/files#diff-a91beeaf44154c199a11f5539f2358704a1df966ff74ad8f2681b64430843f8cL1642

Reply:
Yes, that's agreeable. My concern wasn't the spec and just a way to exercise the bug in the code previously.

I have slight concerns if MSC2716 stuff is still working after all of these changes or should be working at all according to the spec. But if the tests still pass, I guess we're still good.